pkg/scaffold/olm-catalog/csv.go

```diff
@@ -0,0 +1,326 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package catalog
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"unicode"
+
+	"github.com/operator-framework/operator-sdk/internal/util/yamlutil"
+	"github.com/operator-framework/operator-sdk/pkg/scaffold"
+	"github.com/operator-framework/operator-sdk/pkg/scaffold/input"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/ghodss/yaml"
+	olmApi "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	CsvYamlFilePrefix = ".csv.yaml"
+	CsvConfigYamlFile = "csv-config.yaml"
+)
+
+type Csv struct {
+	input.Input
+
+	// DeployDir is the dir the SDK should search for deploy files, ex. *crd.yaml.
+	DeployDir string
+	// ConfigFilePath is the location of a configuration file path for this
+	// projects' CSV file.
+	ConfigFilePath string
+	// CsvVersion is the CSV (and operators') current version.
+	CsvVersion string
+}
+
+func (s *Csv) GetInput() (input.Input, error) {
+	if s.Path == "" {
+		fileName := strings.ToLower(s.ProjectName) + CsvYamlFilePrefix
+		s.Path = filepath.Join(scaffold.OlmCatalogDir, fileName)
+	}
+	if s.ConfigFilePath == "" {
+		s.ConfigFilePath = filepath.Join(scaffold.OlmCatalogDir, CsvConfigYamlFile)
+	}
+	if s.DeployDir == "" {
+		s.DeployDir = scaffold.DeployDir
+	}
+	return s.Input, nil
+}
+
+// CustomRender allows a Csv to be written by marshalling
+// olmApi.ClusterServiceVersion instead of writing to a template.
+func (s *Csv) CustomRender() ([]byte, error) {
+	// Get current CSV to update.
+	currCSV, exists, err := getCurrentCSVIfExists(s.Path)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		currCSV = new(olmApi.ClusterServiceVersion)
+		s.initCSVFields(currCSV)
+	}
+
+	csvConfig, err := getCSVConfig(s.ConfigFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = s.updateCSVVersions(currCSV); err != nil {
+		return nil, err
+	}
+	if err = s.updateCSVFromManifestFiles(currCSV, csvConfig); err != nil {
+		return nil, err
+	}
+
+	// A new csv won't have several required fields populated.
+	if err = checkRequiredCSVFields(currCSV); err != nil {
+		if exists {
+			log.Warnf("required csv fields not filled in file %s:%s\n", s.Path, err)
+		} else {
+			// Report required fields to user informationally.
+			log.Infof("fill in the following required fields in file %s:%s\n", s.Path, err)
```

Comment:
Nit to improve readability:
```suggestion
			log.Infof("in the %s file, fill in the following required fields: %s\n", s.Path, err)
```

Reply:
Why is one a warning and one info, as both check for required fields both should have log level info?

Reply:
If a CSV already exists and the user hasn't filled in required fields, they should be pushed to fill those fields with a warning. On CSV initialization they should just be informed.